netty/src/main/java/io/grpc/netty/NettyServer.java

```diff
@@ -258,51 +292,88 @@ public void operationComplete(ChannelFuture future) throws Exception {
         ch.closeFuture().addListener(loopReleaser);
       }
     });
-    // Bind and start to accept incoming connections.
-    ChannelFuture future = b.bind(address);
-    // We'd love to observe interruption, but if interrupted we will need to close the channel,
-    // which itself would need an await() to guarantee the port is not used when the method returns.
-    // See #6850
-    future.awaitUninterruptibly();
-    if (!future.isSuccess()) {
-      throw new IOException(String.format("Failed to bind to address %s", address), future.cause());
+
+    Future<Map<ChannelFuture, SocketAddress>> bindCallFuture =
+        bossExecutor.submit(new Callable<Map<ChannelFuture, SocketAddress>>() {
+          @Override
+          public Map<ChannelFuture, SocketAddress> call() {
+            Map<ChannelFuture, SocketAddress> map = new HashMap<>();
+            for (SocketAddress address: addresses) {
+                ChannelFuture future = b.bind(address);
+                map.put(future, address);
+            }
+            return map;
+          }
+        }
+    );
+    Map<ChannelFuture, SocketAddress> channelFutures;
+    try {
+      Future<Map<ChannelFuture, SocketAddress>> bindCallResult =
+          bindCallFuture.awaitUninterruptibly();
+      if (!bindCallResult.isSuccess()) {
+        throw bindCallResult.cause();
+      }
+      channelFutures = bindCallResult.get();
+    } catch (Throwable ex) {
+      throw new IOException(String.format("Failed to bind to addresses %s",
+          addresses), ex);
     }
-    channel = future.channel();
-    channel.eventLoop().execute(new Runnable() {
-      @Override
-      public void run() {
-        listenSocketStats = new ListenSocket(channel);
-        channelz.addListenSocket(listenSocketStats);
+
+    for (Map.Entry<ChannelFuture, SocketAddress> channelFutureEntry: channelFutures.entrySet()) {
+      // We'd love to observe interruption, but if interrupted we will need to close the channel,
+      // which itself would need an await() to guarantee the port is not used when the method
+      // returns. See #6850
+      ChannelFuture future  = channelFutureEntry.getKey().awaitUninterruptibly();
+      if (!future.isSuccess()) {
+        throw new IOException(String.format("Failed to bind to address %s",
```

Comment:
This leaks all the other ports. When any one bind fails, we should close all the channels. Note that the channelz registration may also be impacted.

Channelz is probably easiest handled by two listeners added immediately after bind. A listener on the bindFuture would register the new socket. A listener on bindFuture.channel().closeFuture() would deregister the socket. I'll note that while closeFuture() may look really useful, you have to be careful with it, and it runs at an awkward time. I suggest only using it when the execution time doesn't matter. closeFuture() is different than "the future returned from close()".

Reply:
It looks that channel will be closed by b.bind() if that fails and channel only be set once b.bind() future isSuccess(). So I think we can trust the result of the future instead of adding listener to clean up by ourselves. Now I collect all the successful channels and then close them all if one fails to avoid leakage. Maybe we should consider allowing partial failures in the future.